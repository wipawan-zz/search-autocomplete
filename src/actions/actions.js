// action types
const CHANGE_QUERY = "CHANGE_QUERY";
const FETCH_REPOS = "FETCH_REPOS";
const CLEAR_AUTOCOMPLETE = 'CLEAR_AUTOCOMPLETE';
const SELECT_ITEM = 'SELECT_ITEM';

export function changeQuery(query) {
  return {
    type: CHANGE_QUERY,
    query: query
  };
}

export function fetchRepos() {
  return {
    type: FETCH_REPOS
  };
}

export function clearSearch() {
  return {
    type: CLEAR_AUTOCOMPLETE
  }
}

export function selectItem(item) {
  return {
    type: SELECT_ITEM,
    selectedItem: item
  }
}