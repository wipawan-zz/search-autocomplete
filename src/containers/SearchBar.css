.search-container {
  width: 500px;
  height: 200px;
  margin: auto;
}

.search-bar {
  position: relative;
}

.search-input {
  width: 100%;
  margin: 0;
  padding: 12px;
  text-indent: 1em;
  box-sizing: border-box;
  border: 1px solid #ccc;
  border-radius: 4px;
  font-size: 16px;
  line-height: 0.5em;
  outline: none;
}

.search-input:focus {
  border: 1px solid rgb(114, 114, 114);
}


.search-autocomplete {
  /* opacity: 0; */
  visibility: hidden;
  height: 100%;
  list-style-type: none;
  padding: 0;
  margin: 0;
  text-align: left;
  overflow-y: auto;
  overflow-x: hidden;
  z-index: 1;
  background-color: white;
  border: 1px solid #ccc;
  border-top: none;
}

.search-autocomplete-item {
  width: 100%;
  font-size: 14px;
  line-height: 2em;
  text-indent: 1em;
}

.search-autocomplete-item:hover {
  background-color: rgb(224, 224, 224);
}