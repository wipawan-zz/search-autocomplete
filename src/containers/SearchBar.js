import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as Actions from '../actions/actions';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import './SearchBar.css';

const searchIcon = {
  position: "absolute",
  left: "8px",
  top: "14px",
  color: "#ccc"
}

class SearchBar extends Component {
  constructor(props){
    super(props);
    this.state = {
      q: "",
      show: {
        visibility: "visible"
      }
    };
  }

  onChangeQuery = event => {
    this.setState({
      q: event.target.value
    }, () => {
      this.props.changeQuery(this.state.q);
      if(this.state.q.length > 0) {
        this.props.fetchRepos();
        this.setState({
          show: {
            visibility: "visible"
          }
        });
      } else {
        this.props.clearSearch();
      }
    });
  }

  onBlur = () => {
    this.setState({
      show: {
        visibility: "hidden"
      }
    });
  }

  render() {
    const {
      results,
      selectItem
    } = this.props;

    return (
      <div className="search-container">
        <div className="search-bar">
          <input
            className="search-input"
            placeholder="Type your search query"
            type="text"
            value={this.state.q}
            onChange={this.onChangeQuery}
            onBlur={this.onBlur}
          />
          <FontAwesomeIcon style={searchIcon} icon="search" />
        </div>
        <ul className="search-autocomplete" style={results.length > 0 ? this.state.show : {}}>
        {
          results.map((item, index) => <li className="search-autocomplete-item" key={index} onMouseDown={() => selectItem(item)}>{item.full_name}</li>)
        }
        </ul>
      </div>
    )
  }
}

function mapStateToProps(state) {
  return { results: state.results };
}

function mapActionCreatorsToProps(dispatch) {
  return bindActionCreators(Actions, dispatch);
}

export default connect(mapStateToProps, mapActionCreatorsToProps)(SearchBar);