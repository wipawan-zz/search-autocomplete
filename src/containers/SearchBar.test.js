import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';

import SearchBar from './SearchBar';
import { stateMock } from '../stateMock';
import * as Actions from '../actions/actions';

const mockStore = configureStore();
const store = mockStore(stateMock);

describe('<SearchBar />', () => {
  describe('render()', () => {
    test('renders the component', () => {
      const wrapper = shallow(<SearchBar store={store} />);
      const component = wrapper.dive();

      expect(component).toHaveLength(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    test('hides autocomplete box by default', () => {
      const wrapper = shallow(<SearchBar store={store} />);
      const component = wrapper.dive();

      expect(component.find('ul').getElement().props.style.visibility).toEqual('hidden');
    });
  });

  describe('onFocus', () => {
    test('shows autocomplete box when search bar is in focus', () => {
      const wrapper = shallow(<SearchBar store={store} />);
      const component = wrapper.dive();

      // by default
      expect(component.find('ul').getElement().props.style.visibility).toEqual('hidden');

      component.find('input').simulate('focus');

      expect(component.find('ul').getElement().props.style.visibility).toEqual('visible');
    });
  });

  describe('onBlur', () => {
    test('hides autocomplete box when search bar is not in focus', () => {
      const wrapper = shallow(<SearchBar store={store} />);
      const component = wrapper.dive();

      // by default
      expect(component.find('ul').getElement().props.style.visibility).toEqual('hidden');

      component.find('input').simulate('blur');

      expect(component.find('ul').getElement().props.style.visibility).toEqual('hidden');
    });

    test('hides autocomplete box when search bar is empty', () => {
      const wrapper = shallow(<SearchBar store={store} />);
      const component = wrapper.dive();

      // by default
      expect(component.find('ul').getElement().props.style.visibility).toEqual('hidden');

      component.find('input').simulate('focus');

      expect(component.find('ul').getElement().props.style.visibility).toEqual('visible');

      store.dispatch(Actions.clearSearch());

      expect(wrapper.dive().find('ul').getElement().props.style.visibility).toEqual('hidden');
    });
  });
});