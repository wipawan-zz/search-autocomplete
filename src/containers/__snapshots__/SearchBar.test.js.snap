// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> render() renders the component 1`] = `
<div
  className="search-container"
>
  <div
    className="search-bar"
  >
    <input
      className="search-input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Type your search query"
      type="text"
      value=""
    />
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon="search"
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      style={
        Object {
          "color": "#ccc",
          "left": "8px",
          "position": "absolute",
          "top": "14px",
        }
      }
      symbol={false}
      transform={null}
    />
  </div>
  <ul
    className="search-autocomplete"
    style={
      Object {
        "visibility": "hidden",
      }
    }
  >
    <li
      className="search-autocomplete-item"
      key="0"
      onMouseDown={[Function]}
    >
      d3/d3
    </li>
    <li
      className="search-autocomplete-item"
      key="1"
      onMouseDown={[Function]}
    >
      dlang/dmd
    </li>
    <li
      className="search-autocomplete-item"
      key="2"
      onMouseDown={[Function]}
    >
      dlang/druntime
    </li>
    <li
      className="search-autocomplete-item"
      key="3"
      onMouseDown={[Function]}
    >
      vibe-d/vibe.d
    </li>
    <li
      className="search-autocomplete-item"
      key="4"
      onMouseDown={[Function]}
    >
      sjl/d
    </li>
    <li
      className="search-autocomplete-item"
      key="5"
      onMouseDown={[Function]}
    >
      stolksdorf/homebrewery
    </li>
    <li
      className="search-autocomplete-item"
      key="6"
      onMouseDown={[Function]}
    >
      ChaosForge/doomrl
    </li>
    <li
      className="search-autocomplete-item"
      key="7"
      onMouseDown={[Function]}
    >
      felixendres/rgbdslam_v2
    </li>
    <li
      className="search-autocomplete-item"
      key="8"
      onMouseDown={[Function]}
    >
      dlang/dub
    </li>
    <li
      className="search-autocomplete-item"
      key="9"
      onMouseDown={[Function]}
    >
      ldc-developers/ldc
    </li>
    <li
      className="search-autocomplete-item"
      key="10"
      onMouseDown={[Function]}
    >
      Circular-Studios/Dash
    </li>
    <li
      className="search-autocomplete-item"
      key="11"
      onMouseDown={[Function]}
    >
      evanbergeron/DND-5e-LaTeX-Template
    </li>
    <li
      className="search-autocomplete-item"
      key="12"
      onMouseDown={[Function]}
    >
      ethz-asl/libpointmatcher
    </li>
    <li
      className="search-autocomplete-item"
      key="13"
      onMouseDown={[Function]}
    >
      purcell/emacs.d
    </li>
    <li
      className="search-autocomplete-item"
      key="14"
      onMouseDown={[Function]}
    >
      dlang/tools
    </li>
    <li
      className="search-autocomplete-item"
      key="15"
      onMouseDown={[Function]}
    >
      morepurplemorebetter/MPMBs-Character-Record-Sheet
    </li>
    <li
      className="search-autocomplete-item"
      key="16"
      onMouseDown={[Function]}
    >
      dlang/phobos
    </li>
    <li
      className="search-autocomplete-item"
      key="17"
      onMouseDown={[Function]}
    >
      SamPutnam/Index-2018
    </li>
    <li
      className="search-autocomplete-item"
      key="18"
      onMouseDown={[Function]}
    >
      dlang/visuald
    </li>
    <li
      className="search-autocomplete-item"
      key="19"
      onMouseDown={[Function]}
    >
      crobi/dnd5e-quickref
    </li>
    <li
      className="search-autocomplete-item"
      key="20"
      onMouseDown={[Function]}
    >
      D-Programming-GDC/GDC
    </li>
    <li
      className="search-autocomplete-item"
      key="21"
      onMouseDown={[Function]}
    >
      Valloric/statblock5e
    </li>
    <li
      className="search-autocomplete-item"
      key="22"
      onMouseDown={[Function]}
    >
      json-schema-org/json-schema-spec
    </li>
    <li
      className="search-autocomplete-item"
      key="23"
      onMouseDown={[Function]}
    >
      raulmur/ORB_SLAM2
    </li>
    <li
      className="search-autocomplete-item"
      key="24"
      onMouseDown={[Function]}
    >
      cocos2d/cocos2d-x
    </li>
    <li
      className="search-autocomplete-item"
      key="25"
      onMouseDown={[Function]}
    >
      msgpack/msgpack-d
    </li>
    <li
      className="search-autocomplete-item"
      key="26"
      onMouseDown={[Function]}
    >
      blebson/DLink-Camera-Manager
    </li>
    <li
      className="search-autocomplete-item"
      key="27"
      onMouseDown={[Function]}
    >
      ubilabs/kd-tree-javascript
    </li>
    <li
      className="search-autocomplete-item"
      key="28"
      onMouseDown={[Function]}
    >
      Chuyu-Team/Dism-Multi-language
    </li>
    <li
      className="search-autocomplete-item"
      key="29"
      onMouseDown={[Function]}
    >
      pydata/xarray
    </li>
  </ul>
</div>
`;
